### ISSUE TITLE
Vehicle water tanks

### ISSUE THREAD

**Issue by SwagVozhd** [CleverRaven/Cataclysm-DDA#24991]
**Describe the bug**  
Cannot fill a vehicle tank with water via a plastic jerrycan

**To Reproduce**  
1. Equip player with rubber hose and plastic jerrycan.
2. Fill jerrycan with water from a water source.
3. (e)xamine a vehicle with an empty vehicle tank.
4. Re(f)ill tank. 
5. Receive message "You don't have the necessary items at hand"

Alternatively

1. Equip player with rubber hose and plastic jerrycan.
2. Fill jerrycan with water from a water source.
3. (U)nload plastic jerrycan
4. Receive dialogue box with no option to fill the vehicle

**Expected behavior**  
To be able to re(f)ill a vehicle tank with water from a plastic jerrycan. 
Because you can re(f)ill a vehicle tank with gasoline from a plastic  jerrycan, therefore there should be no difference.

**Screenshots**  
Inventory configuration

![screenshot 3](https://user-images.githubusercontent.com/42531157/44322713-f73d4100-a491-11e8-99bb-f710f736020f.png)

After attempting to re(f)ill tank

![screenshot 2](https://user-images.githubusercontent.com/42531157/44322719-fc9a8b80-a491-11e8-9e42-65dd74b4a66d.png)

After attempting to (U)nload plastic jerrycan

![screenshot 1](https://user-images.githubusercontent.com/42531157/44322721-fefce580-a491-11e8-9ade-26cb5522ef80.png)

**Versions and configuration**
 - OS: Windows 10
 - Game Version: 0.C-30730-g4101bda
 - Graphics version: Tiles
 - Mods loaded: DDA, craftable gun pack, bright nights, bionics systems mod, deadleaves' fictional guns, extended realistic guns, generic guns, icecoon's arsenal, makeshift items mod, medieval and historic content, more survival tools, alternative map key, more buildings, more locations, tall buildings, urban development, boats, folding parts pack, vehicle additions pack, tanks and other vehicles, statsthroughskills, simplified nutrition, prevent zombie revivification, classes and scenarios mod.

**Comment by SwagVozhd** [comment]
This issue is exactly the same with clean water.

**Comment by ZhilkinSerg** [comment]
What's with other liquids?

**Comment by SwagVozhd** [comment]
@ZhilkinSerg both gasoline and diesel work fine, I haven't tried any other liquids other than water and clean water though.

**Comment by nexusmrsep** [comment]
Related: #24946

**Comment by Qrox** [comment]
Could it be that liquids with different temperatures do not stack? So with the recent introduction of food temperature, if water in the tank and in the jerrycan have different temperatures this may happen. It's just my guess, haven't tested it.

**Comment by Xelat84** [comment]
Same issue, can't fill vehicle empty tanks with salted water from shallow water.
Also can't fill barrels in nearby tiles when examine shallow water, unti I stand in the same tile with barrel.

**Comment by alanbrady** [comment]
I dug through it all and it seems the root cause is #24228 which now makes all liquids active.

I believe the correct fix involves adding vehicle containers to processing for active liquids.  This then also implies liquids inside containers can freeze.

**Comment by crowder** [comment]
I'd be interested in contributing a fix for this; are there good (or even bad) docs on getting setup to build/debug on Mac OS X?

**Comment by alanbrady** [comment]
Having looked at what the problem is, I definitely wouldn't recommend this as a first time issue.  It involves a lot of moving parts and what looks like right fix isn't actually the right fix.

There's a small hack you can do that makes it so you can put liquids that don't rot into vehicle containers and I think that might suffice as a temporary fix, but what actually needs to happen is vehicle containers need to added to the loop for processing active items and gracefully handling that, that way you can put active items in them.

I thought there were some instructions for compiling on Mac but I couldn't immediately find it.  I think there's some people on discord who know how if it's not in the repo somewhere.

**Comment by AMurkin** [comment]
Don't know is it good: [COMPILING.md#mac-os-x](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/COMPILING.md#mac-os-x)

**Comment by Kevak** [comment]
Tbh this should be high priority as it makes vehicle based strategies an absolute pain in the ass.

**Comment by lispcoc** [comment]
Fill a vehicle tank with active water is restricted because of #18570. But I think this is already fixed by #20010. So I think there is no reason that we still restrict active water.

**Comment by alanbrady** [comment]
#20010 has nothing to do with #18570.  #18570 is talking about invalidating an item_location which we use inappropriately all of the over place.  We have some workarounds for them but it needs a proper solution.  If you just unforbid active items it's likely to cause a segfault in some situation where the item_location we're using gets invalidated.

Additionally, process_vehicle_items doesn't process active items in vehicle tanks, so you need to take care of that as well.

**Comment by alanbrady** [comment]
One more follow-up.  I've found some time to trace through this some more and here's what I've got.

So a veh_interact actually gets 'serialized' (misnomer) into a player_activity here: https://github.com/CleverRaven/Cataclysm-DDA/blob/master/src/veh_interact.cpp#L69

This happens here: https://github.com/CleverRaven/Cataclysm-DDA/blob/419dca391d1f2a8c2f953bfa084454ee15901594/src/game.cpp#L7029

My workaround here: https://github.com/CleverRaven/Cataclysm-DDA/pull/25169 actually covers the potential segfault that can be caused here with processing an active item in a player_activity.

So there's still some issues that need to be taken care of but the segfault is ok.  With that I'm going to try and push a patch up today to close this issue.
